lua/ttt2/libraries/entspawn.lua

```diff
@@ -0,0 +1,235 @@
+	RemoveEntities(map.GetAmmoSpawnEntities(), spawnTable)
+	RemoveEntities(map.GetPlayerSpawnEntities(), spawnTable)
+
+	return spawnTable
+end
+
+---
+-- Spawns weapon and ammo entities on provided spawn point table. The spawn point table already
+-- has the entity types defined.
+-- @param table spawns A table that contains the spawns where entities should be spawned
+-- @param table entsForTypes A table that assigns the ent types to a list of possible entities
+-- @param table entTable A single indexed list that contains all entites without type grouping
+-- @param number randomType The spawn type that should be used as random
+-- @realm server
+function entspawn.SpawnEntities(spawns, entsForTypes, entTable, randomType)
+	for entType, spawnTable in pairs(spawns) do
+		for i = 1, #spawnTable do
+			local spawn = spawnTable[i]
+
+			-- if the weapon spawn is a random weapon spawn, select any spawnable weapon
+			local selectedEnt
+			if entType == randomType then
+				selectedEnt = GetRandomEntityFromTable(entTable)
+			else
+				selectedEnt = GetRandomEntityFromTable(entsForTypes[entType])
+			end
+
+			if not selectedEnt then continue end
+
+			-- create the weapon entity
+			local spawnedEnt = ents.Create(WEPS.GetClass(selectedEnt))
+
+			if not IsValid(spawnedEnt) then continue end
+
+			-- set the position and angle of the spawned weapon entity
+			spawnedEnt:SetPos(spawn.pos)
+			spawnedEnt:SetAngles(spawn.ang)
+			spawnedEnt:Spawn()
+			spawnedEnt:PhysWake()
+
+			-- if the spawn has a defined auto ammo spawn, the ammo should be spawned now
+			if spawn.ammo == 0 or not spawnedEnt.AmmoEnt then continue end
+
+			for k = 1, spawn.ammo do
+				local ammoEnt = ents.Create(spawnedEnt.AmmoEnt)
+
+				if not IsValid(ammoEnt) then continue end
+
+				local pos = spawn.pos + Vector(mathRand(-35, 35), mathRand(-35, 35), 25)
+
+				ammoEnt:SetPos(pos)
+				ammoEnt:SetAngles(VectorRand():Angle())
+				ammoEnt:Spawn()
+				ammoEnt:PhysWake()
+			end
+		end
+	end
+end
+
+---
+-- Spawns all available players.
+-- @param[opt] boolean deadOnly Set to true to only respawn dead players
+-- @realm server
+function entspawn.SpawnPlayers(deadOnly)
+	local waveDelay = cvSpawnWaveInterval:GetFloat()
+	local plys = player.GetAll()
+
+	-- simple method, spawn everybody at once
+	if waveDelay <= 0 or deadOnly then
+		for i = 1, #plys do
+			local ply = plys[i]
+			local spawnPoint = plyspawn.GetRandomSafePlayerSpawnPoint(ply)
+
+			ply:SpawnForRound(deadOnly, spawnPoint.pos, spawnPoint.ang)
+		end
+	else
+		-- wave method
+		local amountSpawns = #plyspawn.GetPlayerSpawnPoints()
+		local toSpawn = {}
+
+		for _, ply in RandomPairs(plys) do
+			if ply:ShouldSpawn() then
+				toSpawn[#toSpawn + 1] = ply
+
+				GAMEMODE:PlayerSpawnAsSpectator(ply)
+			end
+		end
+
+		local sfn = function()
+			local c = 0
+			-- fill the available spawnpoints with players that need spawning
+
+			while c < amountSpawns and #toSpawn > 0 do
```

Comment:
This is legacy TTT code that I just moved in a new function. If I start changing anything there, I completely rewrite it, because it is really shit code. :D

Reply:
I looked over it for a while, the idea is good, but the code is shit. Maybe I'll do a pr for you.

Reply:
It works though. So you should do a PR to the master